File: tests/integration/benchmarks/fixtures/ruby/rosetta/lib/processor.rb
# frozen_string_literal: true

# Core processing logic for the Rosetta benchmark fixture.
#
# This module exercises multiple require patterns:
# - require_relative for local modules
# - Module mixing patterns

require_relative "models"
require_relative "types"
require_relative "helpers"

module Rosetta
  # Processor for batch record operations.
  module Processor
    # Default configuration for processing.
    DEFAULT_CONFIG = ProcessorConfig.new(
      batch_size: 100,
      timeout: 5000,
      strict: true
    ).freeze

    module_function

    # Processes a batch of records and generates a report.
    #
    # @param records [Array<Record>] Records to process
    # @param config [ProcessorConfig, nil] Optional configuration
    # @return [Report] Summary report
    # @raise [ArgumentError] if configuration is invalid
    def run(records, config = nil)
      config ||= DEFAULT_CONFIG

      unless ModelFactory.validate_config(config)
        raise ArgumentError, "Invalid processor configuration"
      end

      values = records.map(&:value)
      total = Helpers.sum_values(values)
      avg = Helpers.average_values(values)

      Report.new(total: total, average: avg, records: records)
    end

    # Creates a formatted summary string from a report.
    #
    # @param report [Report] Report to summarize
    # @return [String] Human-readable summary
    def summarize(report)
      "Total: #{Helpers.format_value(report.total)}, " \
        "Average: #{Helpers.format_value(report.average)}, " \
        "Count: #{report.records.size}"
    end
  end
end

File: tests/integration/benchmarks/fixtures/rust/rosetta/src/processor.rs
//! Core processing logic for the Rosetta benchmark fixture.
//!
//! This module exercises multiple import patterns:
//! - Glob use: `use crate::models::*`
//! - Selective use: `use crate::helpers::{format, sum, average}`
//! - Type imports: `use crate::types::ProcessorConfig`

use crate::models::{self, Record, Report, validate_config};
use crate::types::ProcessorConfig;
use crate::helpers::{format, sum, average};

/// Default configuration for processing.
fn default_config() -> ProcessorConfig {
    ProcessorConfig::new(100, 5000, true)
}

/// Processes a batch of records and generates a report.
///
/// Uses module imports to access Record and Report types,
/// demonstrating how adapters should handle various use patterns.
///
/// # Arguments
/// * `records` - Records to process
/// * `config` - Optional processing configuration
///
/// # Returns
/// Summary report of processed records
///
/// # Panics
/// Panics if configuration is invalid
pub fn run(records: Vec<Record>, config: Option<ProcessorConfig>) -> Report {
    let config = config.unwrap_or_else(default_config);

    if !validate_config(&config) {
        panic!("Invalid processor configuration");
    }

    let values: Vec<f64> = records.iter().map(|r| r.value).collect();
    let total = sum(&values);
    let avg = average(&values);

    Report::new(total, avg, records)
}

/// Creates a formatted summary string from a report.
///
/// # Arguments
/// * `report` - Report to summarize
///
/// # Returns
/// Human-readable summary
pub fn summarize(report: &Report) -> String {
    format!(
        "Total: {}, Average: {}, Count: {}",
        format(report.total),
        format(report.average),
        report.records.len()
    )
}
